[DEV-287] - Change function visibility according to slither suggestions

grp06 opened the pull request:
- Updated Solidity to version 0.8.9 according to Slither's suggestion:
`Pragma version 0.8.9 (src/Raft.sol#2) necessitates a version too recent to be trusted. Consider deploying with 0.6.12/0.7.6/0.8.7`
- change initializer visibility to `public` to give flexibility for future child contracts: https://forum.openzeppelin.com/t/slither-upgradeable-contract-does-not-protect-its-initialize-function/28432
- Organized and cleaned up unit tests.

## Unit test checklist
Cases to test for
- [x] give
	- [x] happy path
	- [x] give with bad token uri
	- [x] give with unauthorized sender
	- [x] give, give again
	- [x] give to self
	- [x] give with tokenId that already exists
- [x] take
	- [x] happy path
	- [x] take with bad token uri
	- [x] take with unauthorized sender
	- [x] take, take again
	- [x] take to self
	- [x] take with tokenId that already exists
 - [x] data holder
	 - [x] set data holder
	 - [x] set data holder as non owner
 - [x] create spec as raft owner
	 - [x] happy path
	 - [x] raft owner isn't message sender
	 - [x] already registered spec
 - [x] unequip
	 - [x] happy  path
	 - [x] non authorized unequip
	 - [x] non-existent tokenId
- [x] ownership
	- [x] transfer ownership
	- [x] transfer ownership as non-owner

rahulrumalla commented on src/Badges.sol:
```diff
@@ -116,26 +116,26 @@ contract Badges is
   }
 
-  function unequip(uint256 tokenId) public virtual override {
+  function unequip(uint256 tokenId) external virtual override {
```
Hmm.. I think some of these functions if marked as external, will not be available for other contracts to call them, no?

grp06 replied:
Other contracts could call it but if another contract inherits from this, that contract would not be able to call it internally. You're right, I'll change these back. We should leave that flexibility.

rahulrumalla commented on src/Badges.t.sol:
```diff
@@ -17,14 +17,6 @@ contract ERC1271Mock {
     pass = pass_;
   }
-
-  function isValidSignature(bytes32 hash, bytes memory signature) public view returns (bytes4) {
```
why are we removing this? I think this is needed to check for signatures originating from contracts (liek multisig).. did we remove this test coverage?

grp06 replied:
It was an unused function. It's also unused in the ERC4973 repo

rahulrumalla commented on src/Raft.sol:
```diff
@@ -29,7 +29,7 @@ contract Raft is ERC721EnumerableUpgradeable, UUPSUpgradeable, OwnableUpgradeabl
     string memory name_,
     string memory symbol_
-  ) external initializer {
+  ) public initializer {
```
what's the general rationale here?

grp06 replied:
> You may also consider changing the initializer to `public` if you want the flexibility of calling that initializer from a child contract, for example.

https://forum.openzeppelin.com/t/slither-upgradeable-contract-does-not-protect-its-initialize-function/28432

rahulrumalla reviewed (commented):
Left some comments. Could you perhaps add all your notes for context in the PR, please? I'm lacking context on most of these to provide feedback.

grp06 commented on src/Badges.sol:
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: GPL-3.0-only
-pragma solidity 0.8.9;
+pragma solidity 0.8.7;
```
Slither says `Pragma version0.8.9 (src/Raft.sol#2) necessitates a version too recent to be trusted. Consider deploying with 0.6.12/0.7.6/0.8.7`

grp06 commented on src/Badges.sol:
```diff
@@ -67,7 +67,7 @@ contract Badges is
     string calldata uri,
     bytes calldata signature
-  ) external virtual returns (uint256) {
+  ) external virtual override returns (uint256) {
```
Here it wasn't compiling without `override`, This function is defined in the interface, so we need `override`. I'm actually not sure how it was compiling before

grp06 commented:
- Updated Solidity to version 0.8.9 to be aligned with the latest version of Solidity that Hardhat is compatible with.
- change initializer visibility to `public` to give flexibility for future child contracts: https://forum.openzeppelin.com/t/slither-upgradeable-contract-does-not-protect-its-initialize-function/28432
- Organized and cleaned up unit tests.

## Unit test checklist
Cases to test for
- [ ] give
	- [x] happy path
	- [x] give with bad token uri
	- [x] give with unauthorized sender
	- [x] give, give again
	- [x] give to self
	- [x] give with tokenId that already exists
- [ ] take
	- [x] happy path
	- [x] take with bad token uri
	- [x] take with unauthorized sender
	- [x] take, take again
	- [x] take to self
	- [x] take with tokenId that already exists
 - [x] data holder
	 - [x] set data holder
	 - [x] set data holder as non owner
 - [x] create spec as raft owner
	 - [x] happy path
	 - [x] raft owner isn't message sender
	 - [x] already registered spec
 - [x] unequip
	 - [x] happy  path
	 - [x] non authorized unequip
	 - [x] non-existent tokenId
- [x] ownership
	- [x] transfer ownership
	- [x] transfer ownership as non-owner

rahulrumalla reviewed (approved):
LGTM!
